Add NetworkConfig to TPU v2 QueuedResource

zeleena opened the pull request:
Add NetworkConfig field and its nested fields. 

```release-note:enhancement
tpuv2: added `network_config` field to `google_tpu_v2_queued_resource` resource
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 112 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12482-old..auto-pr-12482) ( 3 files changed, 550 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 151 insertions(+))



## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_queued_resource` (2 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_queued_resource" "primary" {
  tpu {
    node_spec {
      node {
        network_configs {
          can_ip_forward      = # value needed
          enable_external_ips = # value needed
          network             = # value needed
          queue_count         = # value needed
          subnetwork          = # value needed
        }
      }
    }
  }
}


```

modular-magician commented:
#### Tests analytics
Total tests: 7
Passed tests: 3
Skipped tests: 4
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>tpuv2</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceFullExample
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12482/artifacts/22de38a0-876a-4e6f-9a5b-5542d24e183b/build-log/replaying_test.log)

zeleena commented:
Hi @melinath, it's been about two business days and wanted to see if you could please take a look? I was aiming to merge this change prior to the cut for the next release.

melinath commented on mmv1/products/tpuv2/QueuedResource.yaml:
```diff
@@ -107,3 +118,94 @@ properties:
+                  description: |
+                    Network configurations for the TPU node.
+                  immutable: true
+                  default_from_api: true
```
default_from_api has a couple drawbacks:

- Users won't be able to know accurate information about the end state of the resource (even if it could be known)
- Users can't set the field to its empty/zero value.

I just want to double-check that this field (and the rest marked with default_from_api) don't have [other options](https://googlecloudplatform.github.io/magic-modules/develop/diffs/)?

zeleena replied:
Thanks for highlighting. The [`network` and `subnetwork` fields](https://cloud.google.com/tpu/docs/reference/rest/v2/projects.locations.nodes#Node.NetworkConfig) do get a default value from the API if not set. Consequently, those are subfields of `networkConfig`, which has subfields where some have API defaults and some do not, so I set `networkConfig` to get its default from the API so that its subfields like `network` and `subnetwork` can get these default values. This is also the case for a similar resource TPUv2 VM, [where `networkConfig` has `default_from_api: true`](https://github.com/GoogleCloudPlatform/magic-modules/blob/e831a6f681562d82794af71180e36fc9b568224a/mmv1/products/tpuv2/Vm.yaml#L129) and has been working as expected.

I double checked `queueCount` and that does not get a default value from the TPU API, so I removed `default_from_api` from the `queueCount` subfield here and below.

melinath reviewed (commented):
This seems straightforward; just have one question below. I'll also need to manually run a test since it's skipped in VCR.

melinath commented:
Manual beta test run: https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_MMUPSTREAMTESTS_GOOGLEBETA_PACKAGE_TPUV2/277507?hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 112 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12482-old..auto-pr-12482) ( 3 files changed, 548 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 151 insertions(+))



## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_queued_resource` (2 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_queued_resource" "primary" {
  tpu {
    node_spec {
      node {
        network_configs {
          can_ip_forward      = # value needed
          enable_external_ips = # value needed
          network             = # value needed
          queue_count         = # value needed
          subnetwork          = # value needed
        }
      }
    }
  }
}


```

zeleena commented:
> Manual beta test run: https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_MMUPSTREAMTESTS_GOOGLEBETA_PACKAGE_TPUV2/277507?hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true

Thank you for the manual test run. I'm not sure I have proper permissions to view the results, can you let me know if there's any issues or actions I need to take, or can you let me know the best way to setup my account and get permissions to view the test at this link? When I try to login via GitHub, it says `Authenticated successfully, but failed to create a new user since creating new users by "GitHub.com" login is not allowed`

modular-magician commented:
#### Tests analytics
Total tests: 7
Passed tests: 3
Skipped tests: 4
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>tpuv2</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceFullExample
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12482/artifacts/6c311c20-f820-4065-8a5a-2288182acb23/build-log/replaying_test.log)

melinath commented:
> Thank you for the manual test run. I'm not sure I have proper permissions to view the results, can you let me know if there's any issues or actions I need to take, or can you let me know the best way to setup my account and get permissions to view the test at this link? When I try to login via GitHub, it says `Authenticated successfully, but failed to create a new user since creating new users by "GitHub.com" login is not allowed`

Sorry, we can't grant access to the manual runs due to the platform they run on.

melinath reviewed (changes requested):
Sorry for not catching this sooner - it looks like we're missing tests for the `network_configs` field: https://github.com/GoogleCloudPlatform/magic-modules/pull/12482#issuecomment-2515827989 Could you add them?

melinath commented:
TestAccTpuV2QueuedResource_tpuV2QueuedResourceFullExample passed though and that all looks good!

zeleena commented:
> Sorry, we can't grant access to the manual runs due to the platform they run on.

No worries and sounds good, just wanted to make sure there was no action needed from me at that point!

> Sorry for not catching this sooner - it looks like we're missing tests for the network_configs field: https://github.com/GoogleCloudPlatform/magic-modules/pull/12482#issuecomment-2515827989 Could you add them?

It's a copy of the existing `network_config` field which is tested and conflicts with `network_configs`. Since it's mutually exclusive, I can't add it to the full example test since `network_config` is already defined. I suppose I could create a dedicated test for the array field to work around the mutual exclusivity

> TestAccTpuV2QueuedResource_tpuV2QueuedResourceFullExample passed though and that all looks good!

Great, thanks for running this!

melinath reviewed (changes requested):
yes, a separate test for the networkConfigs field would be great. Even though it's a duplicate, we should have tests to cover it.

melinath commented:
/gcbrun

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 184 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12482-old..auto-pr-12482) ( 3 files changed, 719 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 151 insertions(+))



## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_tpu_v2_queued_resource` (3 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_tpu_v2_queued_resource" "primary" {
  tpu {
    node_spec {
      node {
        network_configs {
          can_ip_forward      = # value needed
          enable_external_ips = # value needed
          network             = # value needed
          queue_count         = # value needed
          subnetwork          = # value needed
        }
      }
    }
  }
}


```

modular-magician commented:
#### Tests analytics
Total tests: 8
Passed tests: 3
Skipped tests: 5
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>tpuv2</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceFullExample
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceNetworkConfigsExample
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12482/artifacts/b99f1204-1710-4563-b4be-289bdfd08f8c/build-log/replaying_test.log)

melinath commented on mmv1/templates/terraform/examples/tpu_v2_queued_resource_network_configs.tf.tmpl:
```diff
@@ -0,0 +1,67 @@
+        runtime_version  = "tpu-vm-tf-2.13.0"
+        accelerator_type = "v2-8"
+
+        network_configs = [
+	  {
+            can_ip_forward      = true
+            enable_external_ips = true
+            network             = google_compute_network.network_0.id
+            subnetwork          = google_compute_subnetwork.subnet_0.id
+	    queue_count         = 32
+          },
+	  {
+            can_ip_forward      = true
+            enable_external_ips = true
+            network             = google_compute_network.network_1.id
+            subnetwork          = google_compute_subnetwork.subnet_1.id
+	    queue_count         = 32
+          }
+	]
```
Missing test detection relies on lists using the repeated notation (instead of the array notation)

```suggestion
        network_configs {
            can_ip_forward      = true
            enable_external_ips = true
            network             = google_compute_network.network_0.id
            subnetwork          = google_compute_subnetwork.subnet_0.id
	    queue_count         = 32
	}
        network_configs {
            can_ip_forward      = true
            enable_external_ips = true
            network             = google_compute_network.network_1.id
            subnetwork          = google_compute_subnetwork.subnet_1.id
	    queue_count         = 32
	}
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 182 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12482-old..auto-pr-12482) ( 3 files changed, 715 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 151 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 8
Passed tests: 3
Skipped tests: 5
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>tpuv2</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceFullExample
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceNetworkConfigsExample
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12482/artifacts/0de78b34-7bd5-4b06-adc3-5d7d67c0951e/build-log/replaying_test.log)

melinath commented:
Manual test run for the new test: https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_MMUPSTREAMTESTS_GOOGLEBETA_PACKAGE_TPUV2/279679?hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true

melinath reviewed (changes requested):
Manual test run failed with:

```

    resource_tpu_v2_queued_resource_generated_test.go:169: Step 1/2 error: Error running apply: exit status 1
        Error: Error creating QueuedResource: googleapi: Error 400: Invalid JSON payload received. Unknown name "networkConfigs" at 'queued_resource.tpu.node_spec[0].node': Cannot find field.
        Details:
        [
          {
            "@type": "type.googleapis.com/google.rpc.BadRequest",
            "fieldViolations": [
              {
                "description": "Invalid JSON payload received. Unknown name \"networkConfigs\" at 'queued_resource.tpu.node_spec[0].node': Cannot find field.",
                "field": "queued_resource.tpu.node_spec[0].node"
              }
            ]
          }
        ]
          with google_tpu_v2_queued_resource.qr,
          on terraform_plugin_test.tf line 2, in resource "google_tpu_v2_queued_resource" "qr":
           2: resource "google_tpu_v2_queued_resource" "qr" {
```

melinath commented on mmv1/products/tpuv2/QueuedResource.yaml:
```diff
@@ -54,6 +54,28 @@ examples:
+    test_env_vars:
+      project: 'PROJECT_NAME'
+    skip_vcr: true
+  - name: 'tpu_v2_queued_resource_network_configs'
+    primary_resource_id: 'qr'
+    min_version: 'beta'
+    vars:
+      qr_name: 'test-qr'
+      tpu_name: 'test-tpu'
+      network_name: 'tpu-net'
+      subnet_name: 'tpu-subnet'
+    test_env_vars:
+      project: 'PROJECT_NAME'
+    skip_vcr: true
```
```suggestion
```

melinath commented on mmv1/products/tpuv2/QueuedResource.yaml:
```diff
@@ -107,3 +129,92 @@ properties:
+                  immutable: true
+                  conflicts:
+                    - network_config
+                  item_type:
+                    type: NestedObject
+                    properties:
+                      - name: 'network'
+                        type: String
+                        description: |
+                          The name of the network for the TPU node. It must be a preexisting Google Compute Engine
+                          network. If none is provided, "default" will be used.
+                        immutable: true
+                        default_from_api: true
+                      - name: 'subnetwork'
+                        type: String
+                        description: |
+                          The name of the subnetwork for the TPU node. It must be a preexisting Google Compute
+                          Engine subnetwork. If none is provided, "default" will be used.
+                        immutable: true
+                        default_from_api: true
+                      - name: 'enableExternalIps'
+                        type: Boolean
+                        description: |
+                          Indicates that external IP addresses would be associated with the TPU workers. If set to
+                          false, the specified subnetwork or network should have Private Google Access enabled.
+                        immutable: true
+                        send_empty_value: true
+                      - name: 'canIpForward'
+                        type: Boolean
+                        description: |
+                          Allows the TPU node to send and receive packets with non-matching destination or source
+                          IPs. This is required if you plan to use the TPU workers to forward routes.
+                        immutable: true
+                        send_empty_value: true
+                      - name: 'queueCount'
+                        type: Integer
+                        description: |
+                          Specifies networking queue count for TPU VM instance's network interface.
+                        required: false
+                        immutable: true
```
```suggestion
```

melinath commented on mmv1/templates/terraform/examples/tpu_v2_queued_resource_network_configs.tf.tmpl:
```diff
@@ -0,0 +1,65 @@
+            network             = google_compute_network.network_1.id
+            subnetwork          = google_compute_subnetwork.subnet_1.id
+	    queue_count         = 32
+	}
+      }
+    }
+  }
+}
+
+resource "google_compute_subnetwork" "subnet_0" {
+  provider = google-beta
+
+  name          = "{{index $.Vars "subnet_name"}}-0"
+  ip_cidr_range = "10.0.0.0/16"
+  region        = "us-central1"
+  network       = google_compute_network.network_0.id
+}
+
+resource "google_compute_network" "network_0" {
+  provider = google-beta
+
+  name                    = "{{index $.Vars "network_name"}}-0"
+  auto_create_subnetworks = false
+}
+
+resource "google_compute_subnetwork" "subnet_1" {
+  provider = google-beta
+
+  name          = "{{index $.Vars "subnet_name"}}-1"
+  ip_cidr_range = "10.1.0.0/16"
+  region        = "us-central1"
+  network       = google_compute_network.network_1.id
+}
+
+resource "google_compute_network" "network_1" {
+  provider = google-beta
+
+  name                    = "{{index $.Vars "network_name"}}-1"
+  auto_create_subnetworks = false
+}
```
```suggestion
```

melinath reviewed (changes requested):
we need to remove the unreleased field

melinath commented on mmv1/products/tpuv2/QueuedResource.yaml:
```diff
@@ -107,3 +129,92 @@ properties:
+                    Network configurations for the TPU node.
+                  immutable: true
+                  default_from_api: true
+                  conflicts:
+                    - network_configs
```
```suggestion
```

melinath reviewed (changes requested):
I'm not able to modify the PR myself, so I don't think it's going possible to get this into a mergeable state today. We may as well wait for the networkConfigs field to be present in the API.

zeleena commented:
Thank you for your patience @melinath, the `networkConfigs` field should be rolled out and present in the API now like I mentioned offline last week. Could we retry a manual test run, now that this field is available?

melinath commented:
/gcbrun

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 182 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12482-old..auto-pr-12482) ( 3 files changed, 715 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 151 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 8
Passed tests: 3
Skipped tests: 5
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>tpuv2</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceFullExample
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceNetworkConfigsExample
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12482/artifacts/d63905e2-a289-4abe-817d-c8eafc59a014/build-log/replaying_test.log)

melinath commented:
Running tests: https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_MMUPSTREAMTESTS_GOOGLEBETA_PACKAGE_TPUV2/285644

zeleena commented:
Apologies, it looks like an older candidate than expected was rolled out last week which caused the field to not be included in the latest rollout. I'll check with my team and ping again once I can confirm the field is rolled out.

melinath reviewed (commented):
ack, marking reviewed in the meantime.

melinath commented:
/gcbrun

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 182 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12482-old..auto-pr-12482) ( 3 files changed, 715 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 151 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 8
Passed tests: 3
Skipped tests: 5
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>tpuv2</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceFullExample
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceNetworkConfigsExample
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12482/artifacts/8527fde2-e2ec-4f25-8b81-a75d372d12ac/build-log/replaying_test.log)

melinath commented:
Running tests: https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_MMUPSTREAMTESTS_GOOGLEBETA_PACKAGE_TPUV2/303239

melinath commented:
Looks like this is still pending rollout

zeleena commented:
Sorry for the delay. The fields likely took awhile to rollout due to code freezes, but the fields are fully rolled out now.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 182 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12482-old..auto-pr-12482) ( 4 files changed, 725 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 151 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 8
Passed tests: 3
Skipped tests: 5
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>tpuv2</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

🔴 Tests were added that are skipped in VCR:
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceFullExample
- TestAccTpuV2QueuedResource_tpuV2QueuedResourceNetworkConfigsExample
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12482/artifacts/4bba7f7a-a91b-4035-8b9a-e24506197129/build-log/replaying_test.log)

melinath commented:
Manual test run: https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_MMUPSTREAMTESTS_GOOGLEBETA_PACKAGE_TPUV2/320936

melinath reviewed (changes requested):
New test failure for TestAccTpuV2QueuedResource_tpuV2QueuedResourceNetworkConfigsExample:

```
 Error: Error creating QueuedResource: googleapi: Error 400: Cloud TPU received a bad request. NetworkConfig and NetworkConfigs fields are exclusive, please provide at most one. [EID: 0x992b93e7f0aa837a]
```

The API request only has NetworkConfigs so this seems like an API error:

```
{
 "name": "tf-test-test-qr9la9k3psli",
 "tpu": {
  "nodeSpec": [
   {
    "node": {
     "acceleratorType": "v2-8",
     "networkConfigs": [
      {
       "canIpForward": true,
       "enableExternalIps": true,
       "network": "projects/ci-test-project-188019/global/networks/tf-test-tpu-net9la9k3psli-0",
       "queueCount": 32,
       "subnetwork": "projects/ci-test-project-188019/regions/us-central1/subnetworks/tf-test-tpu-subnet9la9k3psli-0"
      },
      {
       "canIpForward": true,
       "enableExternalIps": true,
       "network": "projects/ci-test-project-188019/global/networks/tf-test-tpu-net9la9k3psli-1",
       "queueCount": 32,
       "subnetwork": "projects/ci-test-project-188019/regions/us-central1/subnetworks/tf-test-tpu-subnet9la9k3psli-1"
      }
     ],
     "runtimeVersion": "tpu-vm-tf-2.13.0"
    },
    "nodeId": "tf-test-test-tpu9la9k3psli",
    "parent": "projects/ci-test-project-188019/locations/us-central1-c"
   }
  ]
 }
}
```

zeleena commented:
Thanks for the manual test run, that error is unexpected since only the array NetworkConfigs was supplied and not the conflicting singular NetworkConfig.

While the API error is being investigated, I simplified this PR to include just the addition of the singular NetworkConfig and leaving the array NetworkConfigs for a future PR and removed the test for the removed NetworkConfigs as well. I also updated the PR title, description, and release note to match.

Please let me know if this looks good to you and if the tests pass now! Thank you in advance for reviewing.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 80 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12482-old..auto-pr-12482) ( 4 files changed, 338 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12482-old..auto-pr-12482) ( 1 file changed, 74 insertions(+))

melinath commented:
new manual test run: https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_BETA_MMUPSTREAMTESTS_GOOGLEBETA_PACKAGE_TPUV2/323030
